flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobResourceRequirements.java

```diff
@@ -0,0 +1,204 @@
+
+    /**
+     * Write {@link JobResourceRequirements resource requirements} into the configuration of a given
+     * {@link JobGraph}.
+     *
+     * @param jobGraph job graph to write requirements to
+     * @param jobResourceRequirements resource requirements to write
+     * @throws IOException in case we're not able to serialize requirements into the configuration
+     */
+    public static void writeToJobGraph(
+            JobGraph jobGraph, JobResourceRequirements jobResourceRequirements) throws IOException {
+        InstantiationUtil.writeObjectToConfig(
+                jobResourceRequirements,
+                jobGraph.getJobConfiguration(),
+                JOB_RESOURCE_REQUIREMENTS_KEY);
+    }
+
+    /**
+     * Read {@link JobResourceRequirements resource requirements} from the configuration of a given
+     * {@link JobGraph}.
+     *
+     * @param jobGraph job graph to read requirements from
+     * @throws IOException in case we're not able to deserialize requirements from the configuration
+     * @throws ClassNotFoundException in case some deserialized classes are missing on the classpath
+     */
+    public static Optional<JobResourceRequirements> readFromJobGraph(JobGraph jobGraph)
+            throws IOException, ClassNotFoundException {
+        return Optional.ofNullable(
+                InstantiationUtil.readObjectFromConfig(
+                        jobGraph.getJobConfiguration(),
+                        JOB_RESOURCE_REQUIREMENTS_KEY,
+                        JobResourceRequirements.class.getClassLoader()));
+    }
+
+    /**
+     * This method validates that the new job vertex parallelisms are less or equal to the max
+     * parallelism. Moreover, it validates that there are no unknown job vertex ids and that we're
+     * not missing any.
+     *
+     * @param jobResourceRequirements contains the new resources requirements for the job vertices
+     * @param maxParallelismPerVertex allows us to look up maximum possible parallelism for a job
+     *     vertex
+     * @return a list of validation errors
+     */
+    public static List<String> validate(
+            JobResourceRequirements jobResourceRequirements,
+            Map<JobVertexID, Integer> maxParallelismPerVertex) {
+        final List<String> errors = new ArrayList<>();
+        final Set<JobVertexID> missingJobVertexIds =
+                new HashSet<>(maxParallelismPerVertex.keySet());
+        for (JobVertexID jobVertexId : jobResourceRequirements.getJobVertices()) {
+            final Optional<Integer> maybeMaxParallelism =
+                    Optional.ofNullable(maxParallelismPerVertex.get(jobVertexId));
+            if (maybeMaxParallelism.isPresent()) {
+                final JobVertexResourceRequirements.Parallelism requestedParallelism =
+                        jobResourceRequirements.findParallelism(jobVertexId).get();
+                if (requestedParallelism.getLowerBound() > requestedParallelism.getUpperBound()) {
```

Comment:
Shouldn't this be checked in the constructor of `Parallelism`?

And should also probably check that the bounds are not negative.

Reply:
> Shouldn't this be checked in the constructor of Parallelism?

This is a minor flaw in submitting the full code in chunks 🙈 This method will validate the user-provided input in the dispatcher (+ rest layer).

The primary purpose of validating everything at once is to get a list of all errors that we can report back to the REST API user in a meaningful human-readable way.

Good suggestion about checking for the negative values 👍. We just need to take the behavior that we've defined in the FLIP into account:

```
Setting the lower or upper bound to -1 will reset its value to the default setting (1 for the lower bound and parallelism, or maxParallelism for the upper bound).
```

Reply:
Got it, reporting all the errors at once makes sense :+1: